### ISSUE TITLE
Connect Envoy sidecar proxy "cannot create upstream cluster without discovery chain"

### ISSUE THREAD

**Issue by Periclod** [hashicorp/consul#6521]
#### Overview of the Issue

I'm facing an issue with the envoy connect proxy. When I add an upstream to a service, the envoy proxy isn't configured correctly by consul, instead getting a `gRPC config stream closed: 2, cannot create upstream cluster without discovery chain` error message. A similar message `Error handling ADS stream: cannot create upstream cluster without discovery chain` is logged by the agent.

The issue is not present in the built-in proxy. Both were started using the `consul connect proxy|envoy` command, although the issue persists if envoy is started manually with the bootstrapped config.

#### Reproduction Steps

1. Create a cluster with at least 1 server and client
1. Place 2 service defintions on the client(s) with one having the other as an upstream
1. Run envoy with `consul connect envoy -sidecar-for [service]`

### Consul info for both Client and Server

<details>
  <summary>Client config</summary>

```
datacenter = "dc"
data_dir = "/opt/consul/"
connect {
  enabled = true
}
retry_join = ["test-server.lxd"]
ports {
  grpc = 8502
}
```

</details>

<details>
  <summary>Server config</summary>

```
datacenter = "dc"
data_dir = "/opt/consul/"
connect {
  enabled = true
}
retry_join = [ "test-server.lxd" ]
server = true
bootstrap_expect = 1
```

</details>

<details>
  <summary>Service defintions</summary>

```
service {
  id = "web-1"
  name = "web"
  port = 8080
  connect {
    sidecar_service {
      proxy {
        upstreams = [{
            destination_name = "redis"
            local_bind_port = 8128
        }]
      }
    }
  }
}
```

```
service {
  id = "redis-1"
  name = "redis"
  port =  15123
  connect {
    sidecar_service {}
  }
}
```

</details>

### Operating system and Environment details

I'm running consul on LXD Containers. The Host is an Ubuntu Server 18.04 LTS and the Containers are Debian 10. Arch is amd64. I could reproduce the issue on Consul 1.6.0 and 1.6.1 as well as envoy 1.9.1, 1.11.0 and 1.11.1.

### Log Fragments

<details>
  <summary>Envoy Logs</summary>

```
[2019-09-22 15:09:41.201][5337][info][main] [source/server/server.cc:516] starting main dispatch loop
[2019-09-22 15:09:41.204][5337][warning][config] [bazel-out/k8-opt/bin/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:87] gRPC config stream closed: 2, cannot create upstream cluster without discovery chain
[2019-09-22 15:09:41.204][5337][info][upstream] [source/common/upstream/cluster_manager_impl.cc:148] cm init: all clusters initialized
[2019-09-22 15:09:41.204][5337][info][main] [source/server/server.cc:500] all clusters initialized. initializing init manager
[2019-09-22 15:09:41.353][5337][warning][config] [bazel-out/k8-opt/bin/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:87] gRPC config stream closed: 2, cannot create upstream cluster without discovery chain
[2019-09-22 15:09:41.353][5337][info][config] [source/server/listener_manager_impl.cc:761] all dependencies initialized. starting workers
[2019-09-22 15:09:41.590][5337][warning][config] [bazel-out/k8-opt/bin/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:87] gRPC config stream closed: 2, cannot create upstream cluster without discovery chain
[2019-09-22 15:09:43.845][5337][warning][config] [bazel-out/k8-opt/bin/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:87] gRPC config stream closed: 2, cannot create upstream cluster without discovery chain
[2019-09-22 15:09:51.225][5337][warning][config] [bazel-out/k8-opt/bin/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:87] gRPC config stream closed: 2, cannot create upstream cluster without discovery chain

```

</details>

<details>
  <summary>Consul Agent logs</summary>

```
2019/09/22 15:09:41 [DEBUG] http: Request GET /v1/agent/services (1.105611ms) from=127.0.0.1:44378
2019/09/22 15:09:41 [DEBUG] http: Request GET /v1/agent/self (1.811474ms) from=127.0.0.1:44378
2019/09/22 15:09:41 [DEBUG] http: Request GET /v1/agent/service/web-1-sidecar-proxy (266.444µs) from=127.0.0.1:44378
2019/09/22 15:09:41 [WARN] agent: Check "service:redis-1-sidecar-proxy:1" socket connection failed: dial tcp 127.0.0.1:21000: connect: connection refused
2019/09/22 15:09:41 [DEBUG] Error handling ADS stream: cannot create upstream cluster without discovery chain
2019/09/22 15:09:41 [DEBUG] Error handling ADS stream: cannot create upstream cluster without discovery chain
2019/09/22 15:09:41 [DEBUG] Error handling ADS stream: cannot create upstream cluster without discovery chain
2019/09/22 15:09:43 [WARN] agent: Check "service:web-1-sidecar-proxy:1" socket connection failed: dial tcp 127.0.0.1:21001: connect: connection refused
2019/09/22 15:09:43 [DEBUG] Error handling ADS stream: cannot create upstream cluster without discovery chain

```

</details>

**Comment by srix55** [comment]
Can confirm. I upgraded Consul to 1.6.1 with Envoy at 1.11.1 and facing this same issue.

**Comment by jrrdev** [comment]
I can confirm this issue too with Consul 1.6.0 / 1.6.1 and Envoy 1.11.1.

In addition, I noticed that the error is only occurring when services are defined at agent's boot. If they are defined after and the configuration is reloaded by `consul reload`, the proxy can start normally. My Systemd configuration is probably the best way to explain this (dirty) workaround :

<details>
  <summary>/etc/systemd/system/consul.service</summary>

```
[Unit]
Description="HashiCorp Consul - A service mesh solution"
Documentation=https://www.consul.io/

Requires=network-online.target
After=network-online.target

[Service]
Type=notify
User=consul
Group=consul

# Hack to circumvent issue 6521
ExecStartPre=/bin/bash /opt/consul/consul_disable_services.sh
ExecStartPost=/bin/bash /opt/consul/consul_enable_services.sh
ExecStartPost=/usr/local/bin/consul reload

ExecStart=/usr/local/bin/consul agent -config-dir=/etc/consul.d/ -config-dir=/etc/consul.d/services/
ExecReload=/usr/local/bin/consul reload

KillMode=process
Restart=on-failure
LimitNOFILE=65536

[Install]
WantedBy=multi-user.target
```
</details>

<details>
  <summary>/opt/consul/consul_disable_services.sh</summary>

```bash
#!/bin/bash

for f in `/usr/bin/find /etc/consul.d/services/ -name '*.json'`;
do
  /bin/mv "$f" "$f.disabled";
done
```
</details>

<details>
  <summary>/opt/consul/consul_enable_services.sh</summary>

```bash
#!/bin/bash

for f in `/usr/bin/find /etc/consul.d/services/ -name '*.json.disabled'`;
do
  /bin/mv "$f" "${f/.disabled/}";
done
```
</details>

### To reproduce the issue with Docker:

<details>
  <summary>consul-envoy Dockerfile</summary>

```dockerfile
FROM consul:latest
FROM envoyproxy/envoy:v1.11.1
COPY --from=0 /bin/consul /bin/consul
ENTRYPOINT ["consul", "connect", "envoy"]
```
</details>

<details>
<summary> ./servers/consul.hcl </summary>

```hcl
retry_join = ["consul_server"]

connect = {
  enabled = true
}

ports = {
  grpc = 8502
}

server = true
ui = false
```
</details>

<details>
<summary> ./client/consul.hcl </summary>

```hcl
retry_join = ["consul_server"],

bind_addr = "{{ GetAllInterfaces | include \"network\" \"172.18.0.0/24\" | attr \"address\" }}",
client_addr = "{{ GetAllInterfaces | include \"network\" \"172.18.1.0/24\" | attr \"address\" }}",

connect {
  enabled = true
}

ports  {
  grpc = 8502
}

server = false
ui = false

services {
  name = "echo-client"
  port = 8080
  connect {
    sidecar_service {
      proxy {
        upstreams {
          destination_name = "echo"
          local_bind_port = 9191
        }
      }
    }
  }
}
services {
  name = "echo"
  port = 9090
  connect {
    sidecar_service {}
  }
}
```
</details>

<details>
<summary> docker-compose.yml </summary>

```yaml
version: '3.7'

services:
  consul_server:
    image: consul:1.6.1
    command: "agent -bootstrap-expect=1 -config-file /etc/consul/consul.hcl"
    volumes:
      - ./servers/:/etc/consul/:ro
    networks:
      - consul
  consul_client:
    image: consul:1.6.1
    command: "agent -config-file /etc/consul/consul.hcl"
    depends_on:
      - consul_server
    volumes:
      - ./client/:/etc/consul/:ro
    networks:
      - consul
      - app
  sidecar_echo_client:
    image: consul-envoy
    command: "-sidecar-for echo-client"
    depends_on:
      - consul_client
    networks:
      - app
    environment:
      CONSUL_GRPC_ADDR: "consul_client:8502"
      CONSUL_HTTP_ADDR: "consul_client:8500"
  sidecar_echo:
    image: consul-envoy
    command: "-sidecar-for echo"
    depends_on:
      - consul_client
    networks:
      - app
    environment:
      CONSUL_GRPC_ADDR: "consul_client:8502"
      CONSUL_HTTP_ADDR: "consul_client:8500"

networks:
  consul:
    ipam:
      driver: default
      config:
        - subnet: 172.18.0.2/24
  app:
    ipam:
      driver: default
      config:
        - subnet: 172.18.1.2/24
```

**Comment by rboyer** [comment]
This will be fixed in an upcoming 1.6.x point release.

**Comment by nitsh** [comment]
@rboyer I'm seeing this issue too. Could you suggest any work around for this?

**Comment by rboyer** [comment]
Hi @nitsh,

The `master` branch is the work-in-progress branch for the next major release (`1.7.x`) and isn't necessarily ready for general use. The `release/1.6.x` branch is the branch for the next minor release in the 1.6 series and should be perfectly fine to use (as long as you are fine running a prerelease build). This bug was fixed in master and backported for the next upcoming 1.6.x release as well.

It sounds like you are comfortable building from source. If so, try building from `release/1.6.x` instead of `master`.

**Comment by galenwarren** [comment]
Fwiw, I was having the same problem and it was resolved by using Consul v1.5.3 instead of v1.6.1.

@rboyer, are you saying that you would expect this to be fixed in v1.6.2 when that is released? Thanks.
